images/webrtc-stream/cmd/webrtc-stream/main.go

```diff
@@ -342,70 +348,132 @@ func ptr[T any](v T) *T {
+		fatal(err)
 
-		var ogg *oggwriter.OggWriter
-		var encoding AudioEncoding
-		for _, event := range res.Events {
-			var chunk AudioChunk
-			err = json.Unmarshal(event.Payload, &chunk)
-			fatal(err)
+		var b writerseeker.WriterSeeker
+		fatal(wav.Encode(&b, stream, format))
 
-			if ogg == nil {
-				encoding = chunk.Encoding
-				ogg, err = oggwriter.NewWith(&b, chunk.Encoding.SampleRate, chunk.Encoding.Channels)
-				fatal(err)
-			} else if encoding != chunk.Encoding {
-				slog.ErrorContext(ctx, "encoding mismatch", "encoding", chunk.Encoding, "expected", encoding)
-				http.Error(w, fmt.Sprintf("encoding mismatch: %s", err), http.StatusInternalServerError)
-				return
-			}
+		w.Header().Set("Content-Type", "audio/wav")
+		_, err = io.Copy(w, b.Reader())
+		fatal(err)
+	})
+}
 
-			dataPath := s.EventsURL + "/events/" + event.ID.String() + "/data"
-			resp, err := http.Get(dataPath)
-			fatal(err)
-			defer resp.Body.Close()
+func (s *Streamer) handle(w http.ResponseWriter, r *http.Request) (*packetReader, *AudioEncoding, error) {
+	ctx := r.Context()
+
+	trackID := r.PathValue("trackID")
+	audioPath := pathJoin(s.EventPathPrefix, "/tracks/"+trackID+"/audio")
+	audioPath = strings.TrimLeft(audioPath, "/")
+	var segments []Segment
+	if segmentQ := r.URL.Query().Get("segments"); segmentQ != "" {
+		if err := json.Unmarshal([]byte(segmentQ), &segments); err != nil {
+			http.Error(w, fmt.Sprintf("error unmarshalling segments: %s", err), http.StatusBadRequest)
+			panic(http.ErrAbortHandler)
+		}
+	}
+	if len(segments) > 1 {
+		http.Error(w, fmt.Sprintf("max 1 segment supported currently"), http.StatusBadRequest)
+		panic(http.ErrAbortHandler)
+	}
+	minTS := uint32(0)
+	maxTS := uint32(math.MaxUint32)
+	if len(segments) == 1 {
+		minTS = segments[0].Start
+		maxTS = segments[0].End
+	}
 
-			body, err := io.ReadAll(resp.Body)
+	res, err := s.QueryEvents.Call(ctx, QueryEventsInput{
+		Path: audioPath,
+		// Limit: nil,
+		Bias: ptr(-1),
+		// After: nil,
+		// Until: nil,
+	})
+	if err != nil {
+		slog.ErrorContext(ctx, "error querying events", "err", err)
+		http.Error(w, fmt.Sprintf("error querying events: %s", err), http.StatusInternalServerError)
+		panic(http.ErrAbortHandler)
+	}
+	slog.InfoContext(ctx, "query events", "len(events)", len(res.Events), "path", audioPath, "more", res.More)
+
+	var packets packetReader
+	var encoding *AudioEncoding
+	for _, event := range res.Events {
+		var chunk AudioChunk
+		err = json.Unmarshal(event.Payload, &chunk)
+		fatal(err)
+
+		if encoding == nil {
+			encoding = ptr(chunk.Encoding)
 			fatal(err)
+		} else if *encoding != chunk.Encoding {
+			slog.ErrorContext(ctx, "encoding mismatch", "encoding", chunk.Encoding, "expected", encoding)
+			http.Error(w, fmt.Sprintf("encoding mismatch: %#v %#v", chunk.Encoding, encoding), http.StatusInternalServerError)
+			panic(http.ErrAbortHandler)
+		}
+		if chunk.Timestamps[0] > maxTS || chunk.Timestamps[len(chunk.Timestamps)-1] < minTS {
+			continue
+		}
```

Comment:
it's fine for now, but this seems like something best handled in the query. did that not work, or did you just not get to that yet?

Reply:
Yeah, I tried a simple comparison like this, but it doesn't give the desired result:
`ts: [{compare: ">=", value: 937200012}, {compare: "<=", value: 937670412}]`
The comparison does *something*, but >= is always true, and <= is always false no matter the numeric value.
I'm sure the JSON functions could be used, though I didn't see an easy way to handle that with the `compare` criteria.